src/main/kotlin/org/rust/cargo/project/model/impl/CargoProjectImpl.kt

```diff
@@ -216,6 +207,88 @@ open class CargoProjectsServiceImpl(
+            val packagesByRoots = workspace.packages.associateBy { it.rootDirectory }
+            val actualFeatures = features.mapNotNullToSet { f ->
+                packagesByRoots[f.pkg.rootDirectory]?.let { PackageFeature(it, f.name) }
+            }
+            when (newState) {
+                FeatureState.Disabled -> {
+                    // When a user disables a feature, all we have to do is add it into `userDisabledFeatures`.
+                    // The state of dependant features will be inferred in `WorkspaceImpl.inferFeatureState`
+                    for (feature in actualFeatures) {
+                        userDisabledFeatures.setFeatureState(feature, FeatureState.Disabled)
+                    }
+                }
+                FeatureState.Enabled -> {
+                    // But when disables, we should ensure `userDisabledFeatures` state is consistent.
+                    //
+                    // For example, consider such a case:
+                    // (let [x] mean the feature was enabled automatically (by default),
+                    // [d] – disabled because of presence in `userDisabledFeatures`,
+                    // [ ] – disabled automatically because of dependency on [d] feature)
+                    //
+                    // [x] f1 = ["f3"]
+                    // [x] f2 = ["f3"]
+                    // [x] f3 = []
+                    //
+                    // Then a user disables `f3`:
+                    // [ ] f1 = ["f3"]
+                    // [ ] f2 = ["f3"]
+                    // [d] f3 = []
+                    // `f1` and `f2` disabled automatically because they depend on `f3`.
+                    //
+                    // Then a user enables `f1`, and this is our case!
+                    // [x] f1 = ["f3"]
+                    // [d] f2 = ["f3"]
+                    // [x] f3 = []
+                    // To enable `f1`, `f3` should be removed from `userDisabledFeatures` (because `f1`
+                    // depends on `f3`). But if we just remove `f3` from `userDisabledFeatures`, `f2` will
+                    // also become enabled, but we don't want it! So we have to remove `f3` and add `f2`
+
+                    workspace.featureGraph.apply(defaultState = FeatureState.Enabled) {
+                        disableAll(userDisabledFeatures.getDisabledFeatures(workspace.packages))
+                        enableAll(actualFeatures)
+                    }.forEach { (feature, state) ->
+                        if (feature.pkg.origin == PackageOrigin.WORKSPACE) {
+                            userDisabledFeatures.setFeatureState(feature, state)
+                        }
+                    }
+                }
+            }.exhaustive
+        }
+    }
+
+    private fun modifyProjectFeatures(
+        cargoProject: CargoProject,
+        action: (CargoProject, CargoWorkspace, userDisabledFeatures: MutableUserDisabledFeatures) -> Unit
```

Comment:
This action looks overcomplicated. You don't use `CargoProject` of `action` anywhere
Personally, I'd change the type of `action` to `(CargoProject) -> UserDisabledFeatures`

Reply:
Oh, let's leave `action: (CargoWorkspace, MutableUserDisabledFeatures) -> Unit`